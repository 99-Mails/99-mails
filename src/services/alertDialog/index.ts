export * from "./alertDialog";
export * from "./alertDialogAdaptor";
