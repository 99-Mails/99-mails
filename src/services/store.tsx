/* eslint-disable */
import { createContext, useReducer, useEffect } from "react";

const storagePrefix = "disposable_mail_";
const storageKey = storagePrefix + "root";

// type IStoreProviderProps = {
//   children: React.ReactNode[] | React.ReactNode;
// };

// type StoreState = {};

export interface IStoreContext {
  sessionId: string;
  setSessionId: (id: string) => void;
  // address: string[],
  // addAddress: (address: string) => void,
  cleanState: () => void;
}

// enum ActionType {
//   DEFAULT = "DEFAULT",
//   SET_ACCESS_TOKEN = "SET_ACCESS_TOKEN",
//   CLEAR_ACCESS_TOKEN = "CLEAR_ACCESS_TOKEN",
//   SET_EXPIRES_IN = "SET_EXPIRES_IN",
//   CLEAR_EXPIRES_IN = "CLEAR_EXPIRES_IN",
//   SET_PROVIDER_TOKEN = "SET_PROVIDER_TOKEN",
//   CLEAR_PROVIDER_TOKEN = "CLEAR_PROVIDER_TOKEN",
//   SET_REFRESH_TOKEN = "SET_REFRESH_TOKEN",
//   CLEAR_REFRESH_TOKEN = "CLEAR_REFRESH_TOKEN",
//   SET_TOKEN_TYPE = "SET_TOKEN_TYPE",
//   CLEAR_TOKEN_TYPE = "CLEAR_TOKEN_TYPE",
// }

// type Action =
//   | { type: "SET_ACCESS_TOKEN"; payload: { access_token: string } }
//   | { type: "CLEAR_ACCESS_TOKEN" }
//   | { type: "SET_EXPIRES_IN"; payload: { expires_in: number } }
//   | { type: "CLEAR_EXPIRES_IN" }
//   | { type: "SET_PROVIDER_TOKEN"; payload: { provider_token: string } }
//   | { type: "CLEAR_PROVIDER_TOKEN" }
//   | { type: "SET_REFRESH_TOKEN"; payload: { refresh_token: string } }
//   | { type: "CLEAR_REFRESH_TOKEN" }
//   | { type: "SET_TOKEN_TYPE"; payload: { token_type: string } }
//   | { type: "CLEAR_TOKEN_TYPE" };

const initialState = {
  sessionID: null,
  // address: [],
};

const reducer = (state, action) => {
  switch (action.type) {
    case "CLEAN_STATE":
      return initialState;
    case "SET_SESSION_ID":
      return {
        ...state,
        sessionID: action.payload.id,
      };
    // case "ADD_ADDRESS":
    //   return {
    //     ...state,
    //     address: [...state.address, action.payload.address]
    //   }
    default:
      throw new Error("unhandled action type");
  }
};

export const StoreContext = createContext<IStoreContext | null>(null);

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(
    reducer,
    initialState,
    (initial) => JSON.parse(localStorage.getItem(storageKey)) || initial
  );

  useEffect(() => {
    localStorage.setItem(storageKey, JSON.stringify(state));
  }, [state]);

  const value = {
    // address: state.address,
    sessionId: state.sessionID,
    cleanState: () => dispatch({ type: "CLEAN_STATE" }),
    setSessionId: (id: string) =>
      dispatch({ type: "SET_SESSION_ID", payload: { id } }),
    // addAddress: (address: string) => dispatch({ type: "ADD_ADDRESS", payload: { address } }),
  };

  return (
    <StoreContext.Provider value={value}>{children}</StoreContext.Provider>
  );
};

export { StoreProvider };
