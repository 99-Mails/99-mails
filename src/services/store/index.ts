export * from "./store";
export * from "../tempEmailAdaptor";
