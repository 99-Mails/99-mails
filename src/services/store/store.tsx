import { createContext, useReducer, useEffect, Dispatch } from "react";
import {
  TempEmailReducer,
  TempEmailActions,
  TempEmailInitialState,
  tempEmailInitialState,
} from "./reducers/tempEmailReducer";

const storagePrefix = "disposable_mail_";
const storageKey = storagePrefix + "root";

type IStoreProviderProps = {
  children: React.ReactNode[] | React.ReactNode;
};

type ActionTypes = TempEmailActions;

export interface IStoreContext {
  state: InitialStateType;
  dispatch: Dispatch<ActionTypes>;
}

type InitialStateType = {
  tempEmails: TempEmailInitialState;
};

const initialState = {
  tempEmails: tempEmailInitialState,
};

export const StoreContext = createContext<IStoreContext>({
  state: initialState,
  dispatch: () => null,
});

const rootReducer = (
  { tempEmails }: InitialStateType,
  action: ActionTypes
) => ({
  tempEmails: TempEmailReducer(tempEmails, action),
});

const StoreProvider = ({ children }: IStoreProviderProps) => {
  const [state, dispatch] = useReducer(
    rootReducer,
    initialState,
    // @ts-ignore
    (initial) => JSON.parse(localStorage.getItem(storageKey)) || initial
  );

  useEffect(() => {
    localStorage.setItem(storageKey, JSON.stringify(state));
  }, [state]);

  return (
    <StoreContext.Provider value={{ state, dispatch }}>
      {children}
    </StoreContext.Provider>
  );
};

export { StoreProvider };
