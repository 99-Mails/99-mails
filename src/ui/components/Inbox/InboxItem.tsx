import { AddIcon, MinusIcon, DownloadIcon } from "@chakra-ui/icons";
import {
  Flex,
  Card,
  Link,
  Text,
  CardBody,
  CardHeader,
  CardFooter,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
  AccordionItemProps,
  forwardRef,
} from "@chakra-ui/react";
import type { MailSummary } from "@/domains/Mail";
import { FancyDate } from "@/lib/date";
import { memo } from "react";

const InboxHeadItem = ({ text, value }: { text: string; value: string }) => {
  return (
    <Flex gap="4">
      <Text as="b">{text}</Text>
      <Text>{value}</Text>
    </Flex>
  );
};

type InboxItemProps = AccordionItemProps & {
  mail: MailSummary;
};

const InboxItem = memo(forwardRef<InboxItemProps, "div">((props, ref) => {
  const { mail, ...restProps } = props;
  const { headerSubject, toAddr, fromAddr, text, downloadUrl } = mail;

  return (
    <AccordionItem ref={ref} {...restProps} w="full" >
      {({ isExpanded }) => (
        <Card
          boxShadow="none"
          border="1px"
          borderColor="gray.200"
          borderRadius="0"
        >
          <h2>
            <AccordionButton>
              {isExpanded ? (
                <>
                  <CardHeader flex="1" flexDirection="row" textAlign="left">
                    <InboxHeadItem text="Subject:" value={headerSubject} />
                    <InboxHeadItem text="Recipient:" value={toAddr} />
                    <InboxHeadItem text="Sender:" value={fromAddr} />
                    <InboxHeadItem
                      text="Received at:"
                      value={new FancyDate().toISOString()}
                    />
                  </CardHeader>
                  <MinusIcon fontSize="12px" />
                </>
              ) : (
                <>
                  <CardHeader as="span" flex="1" textAlign="left">
                    <Text as="b">{headerSubject}</Text>
                  </CardHeader>
                  <AddIcon fontSize="12px" />
                </>
              )}
            </AccordionButton>
          </h2>
          <AccordionPanel pb={4}>
            <CardBody border="1px" borderColor="gray.300" p="3">
              {text}
            </CardBody>
            <CardFooter>
              <Link color="blue" href={downloadUrl.href} target="_blank">
                <Flex gap="2" alignItems="center">
                  <DownloadIcon />
                  Download Email
                </Flex>
              </Link>
            </CardFooter>
          </AccordionPanel>
        </Card>
      )}
    </AccordionItem>
  );
}));

export { InboxItem, InboxHeadItem };
