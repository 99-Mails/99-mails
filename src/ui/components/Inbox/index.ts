export * from "./Inbox";
export * from "./InboxItem";
