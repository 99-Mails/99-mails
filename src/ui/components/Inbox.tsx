import { AddIcon, MinusIcon, DownloadIcon } from "@chakra-ui/icons";
import {
  Tag,
  Flex,
  Card,
  Link,
  Text,
  Center,
  Heading,
  Spinner,
  CardBody,
  CardHeader,
  CardFooter,
  Accordion,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
} from "@chakra-ui/react";
import { useEmail } from "../../services/api";
import { useStore } from "../../services/storeService";

const InboxHeadItem = ({ text, value }) => {
  return (
    <Flex gap="4">
      <Text as="b">{text}</Text>
      <Text>{value}</Text>
    </Flex>
  );
};

const InboxItem = ({ mail }) => {
  const { headerSubject, text, fromAddr, downloadUrl, rawSize, toAddr } = mail;

  return (
    <AccordionItem>
      {({ isExpanded }) => (
        <Card boxShadow="none" border="1px" borderColor="gray.200">
          <h2>
            <AccordionButton>
              {isExpanded ? (
                <>
                  <CardHeader flex="1" flexDirection="row" textAlign="left">
                    <InboxHeadItem text="Subject:" value={headerSubject} />
                    <InboxHeadItem text="Recipient:" value={toAddr} />
                    <InboxHeadItem text="Sender:" value={fromAddr} />
                    <InboxHeadItem text="Received at:" value={"now"} />
                  </CardHeader>
                  <MinusIcon fontSize="12px" />
                </>
              ) : (
                <>
                  <CardHeader as="span" flex="1" textAlign="left">
                    <Text as="b">{headerSubject}</Text>
                  </CardHeader>
                  <AddIcon fontSize="12px" />
                </>
              )}
            </AccordionButton>
          </h2>
          <AccordionPanel pb={4}>
            <CardBody border="1px" borderColor="gray.300" p="3">
              {text}
            </CardBody>
            <CardFooter>
              <Link color="blue" href={downloadUrl} target="_blank">
                <Flex gap="2" alignItems="center">
                  <DownloadIcon />
                  Download Email
                </Flex>
              </Link>
            </CardFooter>
          </AccordionPanel>
        </Card>
      )}
    </AccordionItem>
  );
};

const InboxContainer = () => {
  const { sessionID } = useStore();

  const { fetchIncomingEmails } = useEmail();
  const { data, loading, error } = fetchIncomingEmails(sessionID);

  return (
    <>
      <Flex flexDirection={"column"} gap="4">
        <Heading size="xs">
          <Flex gap={2} alignItems={"center"}>
            <Text fontSize="3xl">Inbox</Text>
            <Tag bg="gray.500" color="white">
              {data?.session.mails.length}
            </Tag>
          </Flex>
        </Heading>
        {!data?.session.mails.length && (
          <Card bg="gray.100">
            <CardHeader>
              <Heading size="md">No Emails Yet!</Heading>
            </CardHeader>
          </Card>
        )}
        <Accordion allowMultiple>
          {loading && (
            <Center>
              <Spinner />
            </Center>
          )}
          {data &&
            data.session.mails.map((mail, index) => (
              <InboxItem key={index} mail={mail} />
            ))}
        </Accordion>
      </Flex>
    </>
  );
};

export { InboxItem };
export default InboxContainer;
