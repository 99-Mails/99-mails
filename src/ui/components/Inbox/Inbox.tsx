import { AddIcon, MinusIcon, DownloadIcon } from "@chakra-ui/icons";
import {
  Tag,
  Flex,
  Card,
  Link,
  Text,
  Center,
  Heading,
  Spinner,
  CardBody,
  CardHeader,
  CardFooter,
  Accordion,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
} from "@chakra-ui/react";
import { useFetchIncomingEmails } from "../../../services/api";
import { useTempEmail } from "../../../services/tempEmailAdaptor";
import type { Mail } from "../../../domains/Mail";
import ErrorBoundary, { ErrorFallback } from "../ErrorBoundary";

const InboxHeadItem = ({ text, value }: { text: string; value: string }) => {
  return (
    <Flex gap="4">
      <Text as="b">{text}</Text>
      <Text>{value}</Text>
    </Flex>
  );
};

const InboxItem = ({ mail, ...props }: { mail: Mail }) => {
  const { headerSubject, text, fromAddr, downloadUrl, rawSize, toAddr } = mail;

  return (
    <AccordionItem {...props}>
      {({ isExpanded }) => (
        <Card boxShadow="none" border="1px" borderColor="gray.200">
          <h2>
            <AccordionButton>
              {isExpanded ? (
                <>
                  <CardHeader flex="1" flexDirection="row" textAlign="left">
                    <InboxHeadItem text="Subject:" value={headerSubject} />
                    <InboxHeadItem text="Recipient:" value={toAddr} />
                    <InboxHeadItem text="Sender:" value={fromAddr} />
                    <InboxHeadItem text="Received at:" value={"now"} />
                  </CardHeader>
                  <MinusIcon fontSize="12px" />
                </>
              ) : (
                <>
                  <CardHeader as="span" flex="1" textAlign="left">
                    <Text as="b">{headerSubject}</Text>
                  </CardHeader>
                  <AddIcon fontSize="12px" />
                </>
              )}
            </AccordionButton>
          </h2>
          <AccordionPanel pb={4}>
            <CardBody border="1px" borderColor="gray.300" p="3">
              {text}
            </CardBody>
            <CardFooter>
              <Link color="blue" href={downloadUrl.href} target="_blank">
                <Flex gap="2" alignItems="center">
                  <DownloadIcon />
                  Download Email
                </Flex>
              </Link>
            </CardFooter>
          </AccordionPanel>
        </Card>
      )}
    </AccordionItem>
  );
};

const InboxContainer = () => {
  const { sessionID } = useTempEmail();
  const { data, loading, error, refetch } = useFetchIncomingEmails(sessionID);

  return (
    <>
      <ErrorBoundary
        FallbackComponent={ErrorFallback}
        onReset={() => refetch()}
      >
        <Flex flexDirection={"column"} gap="4">
          <Heading size="xs">
            <Flex gap={2} alignItems={"center"}>
              <Text fontSize="3xl">Inbox</Text>
              <Tag bg="gray.500" color="white">
                {data?.session.mails.length}
              </Tag>
            </Flex>
          </Heading>
          {!data?.session.mails.length && (
            <Card bg="gray.100">
              <CardHeader>
                <Heading size="md">No Emails Yet!</Heading>
              </CardHeader>
            </Card>
          )}
          <Accordion allowMultiple>
            {error && (
              <Center>
                <Text>Error!</Text>
              </Center>
            )}
            {loading && (
              <Center>
                <Spinner />
              </Center>
            )}
            {data &&
              data.session.mails.map((mail: Mail, index: number) => (
                <InboxItem key={index} mail={mail} />
              ))}
          </Accordion>
        </Flex>
      </ErrorBoundary>
    </>
  );
};

export { InboxItem, InboxContainer };
