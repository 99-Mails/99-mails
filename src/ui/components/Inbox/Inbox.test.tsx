import { vi, it } from "vitest";
import { IStoreContext } from "@/services/store";
import {
  renderWithContext,
  screen,
  waitForElementToBeRemoved,
} from "@/tests/test-utils";
import { InboxContainer } from "./Inbox";

let storeProps: IStoreContext;

beforeEach(() => {
  storeProps = {
    dispatch: vi.fn,
    state: {
      tempEmails: {
        sessionId: "111111",
      },
    },
  };
});

it("should render correctly", () => {
  const { asFragment } = renderWithContext(<InboxContainer />, { storeProps });
  expect(asFragment()).toMatchSnapshot();
});

it("should display a text when there are no emails", () => {
  renderWithContext(<InboxContainer />, { storeProps });

  expect(screen.getByTestId("inbox-wrapper")).toBeDefined();
  expect(screen.getByTestId("inbox-wrapper")).toHaveTextContent(
    "No Emails Yet!"
  );
});

it("should show zero count when there are no emails", async () => {
  renderWithContext(<InboxContainer />, { storeProps });

  expect(screen.getByTestId("inbox-emails-count")).toBeDefined();

  await waitForElementToBeRemoved(() => screen.queryAllByText("Loading..."));

  expect(screen.getByTestId("inbox-emails-count")).toHaveTextContent("0");
});

it("should show one when there is an email", async () => {
  storeProps.state.tempEmails.sessionId = "123456";

  renderWithContext(<InboxContainer />, { storeProps });

  expect(screen.getByTestId("inbox-emails-count")).toBeDefined();

  await waitForElementToBeRemoved(() => screen.queryAllByText("Loading..."));

  expect(screen.getByTestId("inbox-emails-count")).toHaveTextContent("1");
});

it("should show the correct subject", () => {
  storeProps.state.tempEmails.sessionId = "123456";

  renderWithContext(<InboxContainer />, { storeProps });

  expect(screen.getByTestId("inbox-item")).toBeDefined();
  expect(screen.getByTestId("inbox-item")).toHaveTextContent(
    "Hey, happy Norowz!"
  );
});
