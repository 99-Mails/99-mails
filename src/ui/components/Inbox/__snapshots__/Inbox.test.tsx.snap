// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1kkt86i"
    data-testid="inbox-wrapper"
  >
    <h2
      class="chakra-heading css-0"
    >
      <div
        class="css-r2hvuq"
      >
        <p
          class="chakra-text css-1tapsqs"
        >
          Inbox
        </p>
        <span
          class="css-x780zy"
          data-testid="inbox-emails-count"
        />
      </div>
    </h2>
    <div
      class="chakra-card css-1t31e1n"
    >
      <div
        class="chakra-card__header css-0"
      >
        <h2
          class="chakra-heading css-0"
        >
          No Emails Yet!
        </h2>
      </div>
    </div>
    <div
      class="chakra-accordion css-0"
      data-testid="inbox-body"
    >
      <div
        class="css-gmuwbf"
      >
        <div
          class="chakra-spinner css-1y7joxr"
        >
          <span
            class="css-8b45rq"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
